### PR TITLE
Adds AssociationListener to UriTransformingEnricher

### PR DESCRIPTION
**Pull request by nakomis**
This ensures that the enricher will publish the transformed port
after the port forwarding rule has been created

Dependant upon https://github.com/apache/incubator-brooklyn/pull/538

### CODE REVIEW

File: portforwarding/src/main/java/brooklyn/networking/subnet/SubnetEnrichers.java [hunk 1]
```diff
@@ -85,6 +92,27 @@ public void init() {
             setConfig(TRANSFORMATION_FROM_EVENT, new UriTransformingFunction(subnetTier));
             setConfig(SUPPRESS_DUPLICATES, true);
         }
+
+        @Override
+        public void setEntity(EntityLocal entity) {
+            super.setEntity(entity);
+            checkArgument(getConfig(SOURCE_SENSOR) instanceof AttributeSensor, "expected SOURCE_SENSOR to be AttributeSensor, found " + getConfig(SOURCE_SENSOR));
```
Comment (aledsage) [inline portforwarding/src/main/java/brooklyn/networking/subnet/SubnetEnrichers.java]: Use the `%s` feature of `checkArgument`, to prevent concatenating strings unless it's really necessary. i.e. 

    checkArgument(getConfig(SOURCE_SENSOR) instanceof AttributeSensor, "expected SOURCE_SENSOR to be AttributeSensor, found %s", getConfig(SOURCE_SENSOR));

File: portforwarding/src/main/java/brooklyn/networking/subnet/SubnetEnrichers.java [hunk 2]
```diff
@@ -85,6 +92,27 @@ public void init() {
...
+            PortForwardManager.AssociationListener listener = new PortForwardManager.AssociationListener() {
+                public void onEvent(PortForwardManager.AssociationMetadata metadata) {
+                    Maybe<MachineLocation> machine = Machines.findUniqueMachineLocation(ImmutableList.of(metadata.getLocation()));
+                    String sensorVal = producer.getAttribute((AttributeSensor)sourceSensor).toString();
+                    URI uri = URI.create(sensorVal);
+                    int port = uri.getPort();
+                    if (sensorVal != null && machine.isPresent() && port != -1) {
```
Comment (aledsage) [inline portforwarding/src/main/java/brooklyn/networking/subnet/SubnetEnrichers.java]: Move this null check to before calling `URI.create(sensorVal)`. Otherwise you'll get an NPE

File: portforwarding/src/main/java/brooklyn/networking/subnet/SubnetEnrichers.java [hunk 3]
```diff
@@ -85,6 +92,27 @@ public void init() {
...
+                        if (metadata.getLocation().equals(machine.get()) && metadata.getPrivatePort() == port) {
+                            log.debug("emitting event {} on {} due to port association {}", new Object[] {sourceSensor, producer, metadata});
```
Comment (aledsage) [inline portforwarding/src/main/java/brooklyn/networking/subnet/SubnetEnrichers.java]: Can we give more context in the log message - imagine being sent a log with this message in. Does the message give enough info to tell you which enricher it relates to, and what it's for (ideally without having to search the code for the message)?

How about:

    log.debug("Simulating sensor-event on new port-association {}, to trigger URI transformation by {}", new Object[] {metadata, UriTransformingEnricher.this});

File: portforwarding/src/main/java/brooklyn/networking/subnet/SubnetEnrichers.java [hunk 4]
```diff
@@ -85,6 +92,27 @@ public void init() {
...
+                            UriTransformingEnricher.this.onEvent(new BasicSensorEvent<Object>(sourceSensor, producer, producer.getAttribute((AttributeSensor)sourceSensor)));
```
Comment (aledsage) [inline portforwarding/src/main/java/brooklyn/networking/subnet/SubnetEnrichers.java]: You already have `sensorVal` variable, so don't need to look up attribute again.

Comment (aledsage) [inline portforwarding/src/main/java/brooklyn/networking/subnet/SubnetEnrichers.java]: I have small concern about threading: this is coming from a different thread than the normal sensor events, which could lead to a race if the attribute is changing at the same time (the attribute-change will be handled in a different potentially-concurrent thread). However I don't think that's going to cause any harm, so happy to ignore.

File: portforwarding/src/test/java/brooklyn/networking/subnet/SubnetTierTest.java [hunk 5]
```diff
@@ -158,6 +158,27 @@ public void testUriTransformingEnricher() throws Exception {
     }
 
     @Test
+    public void testUriTransformingEnricherWithDelayedPortMapping() throws Exception {
```
Comment (aledsage) [inline portforwarding/src/test/java/brooklyn/networking/subnet/SubnetTierTest.java]: Can you confirm: did this test fail before you made the fix?

File: vcloud-director-nat-microservice/release-tags.txt [hunk 6]
```diff
@@ -0,0 +1,20 @@
+# Cloudsoft DNAT micro-service
```
Comment (aledsage) [inline vcloud-director-nat-microservice/release-tags.txt]: Can you remove these changes (for release-tags.txt and build-dist.xml) from this commit, and from this PR. If they are wanted (and they do look useful) then can you submit another PR or at least ensure it's a different commit.

File: portforwarding/src/test/java/brooklyn/networking/subnet/SubnetTierTest.java (hunk 5 above)
Comment (nakomis) [inline portforwarding/src/test/java/brooklyn/networking/subnet/SubnetTierTest.java]: Yes

File: portforwarding/src/main/java/brooklyn/networking/subnet/SubnetEnrichers.java [hunk 7]
```diff
@@ -85,6 +92,29 @@ public void init() {
...
+                        if (machine.isPresent() && port != -1) {
+                            if (metadata.getLocation().equals(machine.get()) && metadata.getPrivatePort() == port) {
+                                log.debug("Simulating sensor-event on new port-association {}, to trigger URI transformation by {}", new Object[] {metadata, UriTransformingEnricher.this});
+                                UriTransformingEnricher.this.onEvent(new BasicSensorEvent<Object>(sourceSensor, producer, sensorVal);
+                            }
+                        }
+                    }
+                }
+            };
+            getConfig(SUBNET_TIER).getPortForwardManager().addAssociationListener(listener, Predicates.alwaysTrue());
```
Comment (aledsage) [inline portforwarding/src/main/java/brooklyn/networking/subnet/SubnetEnrichers.java]: Should also override `destroy` to deregister the listener.

**Comment by aledsage** [comment]
Thanks @nakomis - now that apache/incubator-brooklyn#538 is merged, will merge this.
